byzcoin/service.go

```diff
@@ -1719,6 +1812,19 @@ func (s *Service) verifySkipBlock(newID []byte, newSB *skipchain.SkipBlock) bool
 		if len(header.StateChangesHash) != sha256.Size {
 			return errors.New("state changes hash is wrong size")
 		}
+
+		prevBlock := s.skService().GetDB().GetByID(newSB.BackLinkIDs[0])
+		if prevBlock == nil {
+			return errors.New("missing previous block")
+		}
+		prevHeader, err := decodeBlockHeader(prevBlock)
+		if err != nil {
+			return err
+		}
+		if header.Version < prevHeader.Version {
```

Comment:
is there a "limit" for `header.Version`? for instance, whoever calls `createUpgradeVersionBlock` can create their own version. is there a way to stop a malicious node from proposing a block with version 99999 when the current version is 1? will it be accepted? do we need to prevent such a scenario?

Reply:
We discussed that with @Gilthoniel over luch, and it should indeed make sure that the following should hold:

```
prevVersion <= currentVersion <= conodeVersion
```

Reply:
is `conodeVersion` a hard-coded thing in byzcoin? there might be consensus issues if the conodes use different `conodeVersion`s. suppose this expression `prevVersion <= currentVersion <= conodeVersion` fails on half of the conodes but passes on the other half, then we will not have consensus. depending on the deployment, you could say that if consensus fails for this reason, then the conode admins need to make sure they update their conodes.

Reply:
We talked about this with Noemien, and the conclusion was that for starters, it's simplest if the owner of the leader can call a method to send an empty block with the new version. If 2/3+1 of the nodes accept this block by signing the forward-link, all future blocks will be of the new version.

A more elaborate version of this would be to extend the `CollectTxResponse` structure with a field like `ConodeVersion`. Every node would always fill this field, whenever the leader asks to send the latest transactions. Then the leader could decide by himself if he can get a consensus of nodes to sign a new version or not.